### PR TITLE
chore: greater protection for release mode

### PR DESCRIPTION
**Pull request by rjzak**
* Self-signed certificate not permitted in release mode
* Generating a certificate is not permitted in release mode

References https://github.com/profianinc/steward/issues/49

Signed-off-by: Richard Zak <richard@profian.com>

### CODE REVIEW

**Review by rvolosatovs — CHANGES_REQUESTED** [review]
I don't think that solution to "having too much code" can be adding more code.

Why don't we just remove the testing code from our critical production service instead?

If we need this functionality for unit tests, then let's define clear boundaries in our internal API such that testing would be possible without modifying the implementation and be fully self-contained within the test module?

In any case, tests utilizing testing code don't *test* anything. Tests must rely on "real" implementations we run in production as much as possible.

File: src/main.rs [hunk 1]
```diff
@@ -80,6 +85,7 @@ struct Args {
     #[clap(short, long, env = "ROCKET_ADDRESS", default_value = "::")]
     addr: IpAddr,
 
+    #[cfg(debug_assertions)]
     #[clap(short, long, env = "RENDER_EXTERNAL_HOSTNAME")]
```
[the remark this answers is not in the document]
Comment (rjzak) [reply src/main.rs:89]: It's used to generate keys, which we wouldn't do in a production environment.

[the remark this answers is not in the document]
Comment (rjzak) [reply src/main.rs:89]: Additionally, some tests were disabled for release mode (since Nix seems to test in release mode) due to this discovered bug https://github.com/profianinc/steward/issues/73

File: src/ext/mod.rs [hunk 2]
```diff
@@ -28,5 +28,5 @@ pub trait ExtVerifier {
     /// certificate. Returning `Ok(false)` will allow the certification request
     /// to continue, but this particular extension will not be included
     /// in the resulting certificate.
-    fn verify(&self, cri: &CertReqInfo<'_>, ext: &Extension<'_>, dbg: bool) -> Result<bool>;
+    fn verify(&self, cri: &CertReqInfo<'_>, ext: &Extension<'_>, dbg: &bool) -> Result<bool>;
```
[the remark this answers is not in the document]
Comment (rjzak) [reply src/ext/mod.rs:31]: I did that to address complaints about use of a moved value.

**Comment by rjzak** [comment]
> I don't think that solution to "having too much code" can be adding more code.

I don't think I added that much code. I disabled code when compiling for `release`.

> Why don't we just remove the testing code from our critical production service instead?

We need to have some testing (`debug` mode) to ensure things work in the future, since not everyone has SGX or SNP machines. The very reason for the `nil` and `kvm` backends in Enarx.

> If we need this functionality for unit tests, then let's define clear boundaries in our internal API such that testing would be possible without modifying the implementation and be fully self-contained within the test module?

Sure, I can move the logic for `generate()` to the test side. But I still think having some debug functionally as a stand-alone executable beyond unit tests is helpful. The only part would be to ensure not using self-signed certs when compiled for `release`.
 
> In any case, tests utilizing testing code don't _test_ anything, those paths will never be taken in production setting, since those pieces of code will never be compiled in `release` mode. Note, that a mistake here could be _very_ costly, while also being pretty easy to make. Tests must rely on "real" implementations we run in production as much as possible.

Again, the tests ensure the `debug` parts work, when utilized for actual testing. Better attestation reports and public keys can be serialized to ensure better testing https://github.com/profianinc/steward/issues/73 to fully exercise the `release` code.

File: src/ext/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (haraldh) [reply src/ext/mod.rs:31]: strange, it just compiles fine without it

[the remark this answers is not in the document]
Comment (rjzak) [reply src/ext/mod.rs]: I think my IDE was having problems. I reverted back to `bool` and it's fine.

File: src/ext/kvm.rs [hunk 3]
```diff
@@ -28,10 +28,10 @@ impl ExtVerifier for Kvm {
             return Err(anyhow!("invalid kvm extension"));
         }
 
-        if !dbg {
-            return Err(anyhow!("steward not in debug mode"));
-        }
+        #[cfg(not(feature = "insecure"))]
+        return Err(anyhow!("steward not in debug mode"));
```
Comment (rvolosatovs) [inline src/ext/kvm.rs:32]:
```suggestion
        bail!("steward not in insecure mode")
```

**Review by rvolosatovs — COMMENTED** [review]
As discussed a few times already, deferring review to @npmccallum 
I personally think that we should just remove KVM attestation and use real, captured SGX and SEV reports in tests.
If we really want to have KVM attestation available, then that should be guarded by `insecure` feature (as is already done) and `--insecure` runtime flag. Still, I am not happy by how much the implementation is changed to account for this. I find that the prevalent `cfg` attributes make the code harder to reason about and maintain. What's worse, it introduces plethora of new code paths, that actual, real, production service will never take and results in "real" code that should be tested to not get tested.

If we really want to have KVM attestation available, then I think we need to restructure the code in such a way that this feature would not be as intrusive and such that the unit tests would still test "real" code paths and not code paths, which would never be taken in production.

File: src/ext/kvm.rs (hunk 3 above)
Comment (rjzak) [reply src/ext/kvm.rs:32]: Captured SEV & SGX reports added yesterday.
